packages/beacon-node/src/sync/utils/downloadByRoot.ts

```diff
@@ -0,0 +1,589 @@
+  // not all needed blobs were fetched via getBlobs, need to use ReqResp
+  if (blobSidecars.length !== blobMeta.length) {
+    const networkResponse = await fetchBlobsByRoot({
+      network,
+      peerIdStr,
+      blobMeta,
+      indicesInPossession: blobSidecars.map((b) => b.index),
+    });
+    blobSidecars.push(...networkResponse);
+  }
+
+  // responses can be sparse for both types of requests to sort to make sure its in sequential order
+  blobSidecars.sort((a, b) => a.index - b.index);
+
+  await validateBlockBlobSidecars(block.message.slot, blockRoot, blobMeta.length, blobSidecars);
+
+  return blobSidecars;
+}
+
+export async function fetchGetBlobsV1AndBuildSidecars({
+  config,
+  executionEngine,
+  forkName,
+  block,
+  blobMeta,
+}: Pick<
+  FetchByRootAndValidateBlobsProps,
+  "config" | "executionEngine" | "forkName" | "block" | "blobMeta"
+>): Promise<deneb.BlobSidecars> {
+  const blobSidecars: deneb.BlobSidecars = [];
+
+  const enginedResponse = await executionEngine.getBlobs(
+    forkName,
+    blobMeta.map(({versionedHash}) => versionedHash)
+  );
+
+  if (!enginedResponse.length) {
+    return blobSidecars;
+  }
+
+  // response.length should always match blobMeta.length and they should be in the same order
+  for (let i = 0; i < blobMeta.length; i++) {
+    const blobAndProof = enginedResponse[i];
+    if (blobAndProof) {
+      const {blob, proof} = blobAndProof;
+      const index = blobMeta[i].index;
+      const kzgCommitment = block.message.body.blobKzgCommitments[index];
+      const sidecar: deneb.BlobSidecar = {
+        index,
+        blob,
+        kzgProof: proof,
+        kzgCommitment,
+        // TODO(fulu): refactor this to only calculate the root inside these following two functions once
+        kzgCommitmentInclusionProof: computePreFuluKzgCommitmentsInclusionProof(forkName, block.message.body, index),
+        signedBlockHeader: signedBlockToSignedHeader(config, block),
+      };
+      blobSidecars.push(sidecar);
+    }
+  }
+
+  return blobSidecars;
+}
+
+export async function fetchBlobsByRoot({
+  network,
+  peerIdStr,
+  blobMeta,
+  indicesInPossession = [],
+}: Pick<FetchByRootAndValidateBlobsProps, "network" | "peerIdStr" | "blobMeta"> & {
+  indicesInPossession?: number[];
+}): Promise<deneb.BlobSidecars> {
+  const blobsRequest = blobMeta
+    .filter(({index}) => !indicesInPossession.includes(index))
+    .map(({blockRoot, index}) => ({blockRoot, index}));
+  if (!blobsRequest.length) {
+    return [];
+  }
+  return await network.sendBlobSidecarsByRoot(peerIdStr, blobsRequest);
+}
+
+export async function fetchAndValidateColumns({
+  config,
+  network,
+  executionEngine,
+  forkName,
+  peerIdStr,
+  block,
+  blockRoot,
+  columnMeta,
+}: FetchByRootAndValidateColumnsProps): Promise<fulu.DataColumnSidecars> {
+  const slot = block.message.slot;
+  const blobCount = block.message.body.blobKzgCommitments.length;
+  const blobsV2ColumnSidecars = await fetchGetBlobsV2AndBuildSidecars({
```

Comment:
this still call `getBlobsV2()` over a loop of peers, need to merge with #8282 once it's done